### ISSUE TITLE
Psalm should detect correct code path in while(true) cycle

### ISSUE THREAD

**Issue by gennadiylitvinyuk** [vimeo/psalm#5612]
Psalm incorrectly detects code path in `while(true)` cycle.

Minimal example:
```php
<?php
function ask(): bool
{
    while (true) {
        return false;
    }
}
```
This code produces following error, althow no othe ways to exit the cycle are possible:
```
ERROR: InvalidReturnType - 3:17 - Not all code paths of ask end in a return statement, return type bool expected
````

**Comment by gennadiylitvinyuk** [comment]
Sure, [https://psalm.dev/r/801802130d](https://psalm.dev/r/801802130d)

**Comment by gennadiylitvinyuk** [comment]
Another example: https://psalm.dev/r/ac1c7b92fc

**Comment by AndrolGenhald** [comment]
Also, https://psalm.dev/r/702ac3af45 should work, and https://psalm.dev/r/bc56ebdd71 should not.

**Comment by gennadiylitvinyuk** [comment]
May be the same, may be another case: https://psalm.dev/r/483c321cd9
